### PR TITLE
[WIP] Replaced `Open with Default program` with a more descriptive title

### PR DESCRIPTION
**Pull request by agisilaos**
Fixes: #2904

With this change when the user will right click in one of the changed files instead of getting `Open with Default program` as the title of the action instead will get `Open in ${this.selectedExternalEditor}`. This change will bring clarity to the user when it comes to open a changed file.

### CODE REVIEW

**Comment by shiftkey** [comment]
This is only part of the solution, and as we discussed in #2904 the current behaviour of this callback isn't clear. The `openFile` function will use the Electron `shell` API which relies on the OSes file association APIs to open a program:

https://github.com/desktop/desktop/blob/c23b1ae24cd82233e4b0cb3503f8fc621cb370d5/app/src/lib/open-file.ts#L4-L17

**Comment by agisilaos** [comment]
@shiftkey Thanks for your comment above. I was trying to figure that out. So now when I right click in a changed file I have this label on the button: `Open in ${this.selectedExternalEditor}` and that's not right for the reason that instead of `${this.selectedExternalEditor}` should present one of the 3 editors name.

**Comment by agisilaos** [comment]
Hello @shiftkey I have a couple of questions regarding this issue and how I should procceed. With the current implementation when I right click on a changed file I have this menu and it's wrong: 

screen shot 2017-10-09 at 14 34 16

1) How I can make the menu to display the preffered editor?

My first commit for this PR was like this: 
screen shot 2017-10-09 at 14 47 34

And with my 2nd commit  changed it like this:

screen shot 2017-10-09 at 14 47 44

I think my first commit was the correct approach but when I try to run it I get the error that `selectedExternalEditor` is undefined. How I can fix that? I researching about this issue for a couple of hours now but didn't understand how to fix it. Any help/ suggestions on that?

My second question is about the code block that you commented above. What changes need to e implemented there?
https://github.com/desktop/desktop/blob/c23b1ae24cd82233e4b0cb3503f8fc621cb370d5/app/src/lib/open-file.ts#L4-L17

Thank you in advance for all the help and suggestions so far!

File: app/src/lib/git/diff-index.ts [hunk 1]
```diff
@@ -82,7 +82,7 @@ export async function getIndexChanges(
   const args = ['diff-index', '--cached', '--name-status', '--no-renames', '-z']
 
   let result = await git(
-    [...args, 'HEAD', '--'],
```
Comment (shiftkey) [inline app/src/lib/git/diff-index.ts]: Not sure what happened to your merge, but this change is part of an important bugfix #2990 and some methods that were removed in #2991 are now restored.

**Comment by shiftkey** [comment]
> How I can make the menu to display the preffered editor?

That's the string interpolation support in ES6 that let's you use backticks \` to build up a string, rather using than single-quotes `'` and building it up manually. The compiler error is because it can't find that field.

To get this value into the component, React recommends defining some props in the component to represent what you need:

```js
interface IChangedFileProps {
  ...
  readonly selectedExternalEditor: string
  ...
}
```

This means your new string will become:

```js
`Open in {this.props.selectedExternalEditor}`
```

And if you run `npm run compile:dev` you'll see this introduces compiler errors, because you then need to pass in this value wherever a `ChangedFile` component is used. This means passing some more props in (basically repeating the process) until you get to wherever we're using the AppStore component and can get this value.

> My second question is about the code block that you commented above. What changes need to be implemented there?

That's the logic we're using for opening a file using the OS, which we no longer want to do. But if we're going to use the selected external editor that's a different path. The function for that is found [here](https://github.com/desktop/desktop/blob/master/app/src/lib/editors/launch.ts) but it's all based on the assumption that we're opening a folder - I think it'll be fine for these, but just something to be cautious of when testing.

File: app/src/lib/git/diff-index.ts (hunk 1 above)
Comment (agisilaos) [reply app/src/lib/git/diff-index.ts]: Merged the new changes correctly with my latest commit.

**Comment by agisilaos** [comment]
@shiftkey The last couple of days I'm trying to fetch the latest changes from master but I fail consistently. Not sure how to procceed. I want to fetch and new changes first and then continue working on this issue. Any thoughts what maybe is going wrong?

**Comment by j-f1** [comment]
Can you remove the `package-lock.json` file? @shiftkey is working on moving to Yarn in #2864.

**Comment by agisilaos** [comment]
I'm gonna close this and start fresh with a new PR, I didn't made so many changes here and something not working. Will keep this PR into my sigh for reference with all your valuable information about the issue @shiftkey.

**Pull request closed without merging**